plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { TaskRunner } from '@backstage/backend-tasks';
+import { Config } from '@backstage/config';
+import { AwsS3Integration, ScmIntegrations } from '@backstage/integration';
+import {
+  EntityProvider,
+  EntityProviderConnection,
+  LocationSpec,
+  locationSpecToLocationEntity,
+} from '@backstage/plugin-catalog-backend';
+import { AwsCredentials } from '../credentials/AwsCredentials';
+import { readAwsS3Configs } from './config';
+import { AwsS3Config } from './types';
+import { S3 } from 'aws-sdk';
+import { ListObjectsV2Output } from 'aws-sdk/clients/s3';
+import * as uuid from 'uuid';
+import { Logger } from 'winston';
+
+// TODO: event-based updates using S3 events (+ queue like SQS)?
+/**
+ * Provider which discovers catalog files (any name) within an S3 bucket.
+ *
+ * Use `AwsS3EntityProvider.fromConfig(...)` to create instances.
+ *
+ * @public
+ */
+export class AwsS3EntityProvider implements EntityProvider {
+  private readonly logger: Logger;
+  private readonly s3: S3;
+  private readonly scheduleFn: () => Promise<void>;
+  private connection?: EntityProviderConnection;
+
+  static fromConfig(
+    configRoot: Config,
+    options: {
+      logger: Logger;
+      schedule: TaskRunner;
+    },
+  ): AwsS3EntityProvider[] {
+    const providerConfigs = readAwsS3Configs(configRoot);
+
+    // Even though the awsS3 integration allows a config array
+    // there is no *real* support for multiple configs.
+    // Usually, there will be just the integration for the default host.
+    // In case, a config custom endpoint is used, the host from this endpoint
+    // will be extracted and used as host (e.g., localhost when used with LocalStack)
+    // and the default integration will be added as second integration.
+    // In this case, we still want the first one though, but have no means to select it
+    // just from the bucket name (and region).
+    const integration = ScmIntegrations.fromConfig(configRoot).awsS3.list()[0];
```

Comment:
thought for the future: this actually can never be undefined at the current point in time, but if we ever changed the behavior to not add a default entry silently, it might have been wise to be defensive here and throw an error if integration turned out to be undefined

Reply:
good point. I can create a PR to add this.

Reply:
follow-up PR #10697